libtenzir/builtins/operators/from_to_2.cpp

```diff
@@ -0,0 +1,572 @@
+        strip_scheme(inv.args.front(), url->scheme());
+      }
+      if (io_properties.transform_uri) {
+        TRY(auto uri_replacement,
+            io_properties.transform_uri(get_as_located_string(inv.args.front()),
+                                        ctx));
+        TENZIR_TRACE("{} operator: URI replacement size  : {}",
+                     traits::operator_name, uri_replacement.size());
+        TENZIR_ASSERT(not uri_replacement.empty());
+        inv.args.erase(inv.args.begin());
+        inv.args.insert(inv.args.begin(), uri_replacement.begin(),
+                        uri_replacement.end());
+        if (pipeline_argument) {
+          pipeline_argument = try_as<ast::pipeline_expr>(inv.args.back());
+        }
+      }
+    }
+  } else {
+    io_plugin
+      = plugins::find<operator_factory_plugin>(traits::default_io_operator);
+  }
+  const bool has_pipeline_or_events = pipeline_argument or io_properties.events;
+  auto compression_name = std::string_view{};
+  if (not has_pipeline_or_events) {
+    auto file = get_file(*url);
+    if (file.empty()) {
+      diagnostic::error("URL has no segments to deduce the format from")
+        .primary(inv.args.front().get_location())
+        .hint("you can pass a pipeline to handle compression and format")
+        .docs(docs)
+        .emit(ctx);
+      goto post_deduction_reporting;
+    }
+    auto filename_loc = inv.args.front().get_location();
+    if (filename_loc.end - filename_loc.begin == path.size() + 2) {
+      auto file_start = path.find(file);
+      filename_loc.begin += file_start + 1;
+      filename_loc.end -= 1;
+    }
+    auto first_dot = file.find('.');
+    if (first_dot == file.npos) {
+      if (io_properties.default_format == nullptr) {
+        diagnostic::error("did not find extension in `{}`", file)
+          .primary(filename_loc)
+          .hint("you can pass a pipeline to handle compression and format")
+          .emit(ctx);
+      }
+      goto post_deduction_reporting;
+    }
+    auto file_ending = std::string_view{file}.substr(first_dot);
+    // determine compression based on ending
+    auto compression_extension = std::string_view{};
+    std::tie(compression_extension, compression_name)
+      = determine_compression(file_ending);
+    if (not compression_name.empty()) {
+      for (const auto& p : plugins::get<operator_factory_plugin>()) {
+        const auto name = p->name();
+        // TODO These should ultimately be different operators
+        if (name == traits::compression_operator_name) {
+          compression_plugin = p;
+          break;
+        }
+      }
+      TENZIR_ASSERT(compression_plugin);
+    }
+    // determine read operator based on file ending
+    std::tie(rw_plugin, rw_properties) = find_formatter_given<is_loading>(
+      file_ending, path, compression_extension, inv.args.front().get_location(),
+      docs, io_properties.default_format == nullptr, ctx);
+  }
+post_deduction_reporting:
```

Comment:
It looks to me like the `goto` logic here could have been easily avoided, but it's ok.